include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -143,10 +129,360 @@ class FEEvaluationBase
+
+  /**
+   * Stores the subface index of the given face. Usually, this variable takes
+   * the value numbers::invalid_unsigned_int to indicate integration over the
+   * full face, but in case the current physical face has a neighbor that is
+   * more refined, it is a subface and must scale the entries in ShapeInfo
+   * appropriately.
+   */
+  unsigned int subface_index;
+
+  /**
+   * Stores the type of the cell we are currently working with after a call to
+   * reinit(). Valid values are @p cartesian, @p affine and @p general, which
+   * have different implications on how the Jacobian transformations are
+   * stored internally in MappingInfo.
+   */
+  internal::MatrixFreeFunctions::GeometryType cell_type;
+
+  /**
+   * Geometry data that can be generated FEValues on the fly with the
+   * respective constructor.
+   */
+  std::shared_ptr<internal::MatrixFreeFunctions::
+                    MappingDataOnTheFly<dim, Number, VectorizedArrayType>>
+    mapped_geometry;
+};
+
+
+
+/**
+ * This is the base class for the FEEvaluation classes.  This class needs
+ * usually not be called in user code and does not have any public
+ * constructor. The usage is through the class FEEvaluation instead. It
+ * implements a reinit method that is used to set pointers so that operations
+ * on quadrature points can be performed quickly, access functions to vectors
+ * for the FEEvaluationBase::read_dof_values(),
+ * FEEvaluationBase::set_dof_values(), and
+ * FEEvaluationBase::distribute_local_to_global() functions, as well as
+ * methods to access values and gradients of finite element functions. It also
+ * inherits the geometry access functions provided by the class
+ * FEEvaluationMappingRelated.
+ *
+ * This class has five template arguments:
+ *
+ * @tparam dim Dimension in which this class is to be used
+ *
+ * @tparam n_components Number of vector components when solving a system of
+ * PDEs. If the same operation is applied to several components of a PDE (e.g.
+ * a vector Laplace equation), they can be applied simultaneously with one
+ * call (and often more efficiently)
+ *
+ * @tparam Number Number format, usually @p double or @p float
+ *
+ * @tparam is_face Whether the class is used for a cell integrator (with
+ * quadrature dimension the same as the space dimension) or for a face
+ * integrator (with quadrature dimension one less)
+ *
+ * @tparam VectorizedArrayType Type of array to be woked on in a vectorized
+ *                             fashion, defaults to VectorizedArray<Number>
+ *
+ * @note Currently only VectorizedArray<Number, width> is supported as
+ *       VectorizedArrayType.
+ *
+ *
+ * @ingroup matrixfree
+ */
+template <int dim,
+          int n_components_,
+          typename Number,
+          bool is_face                 = false,
+          typename VectorizedArrayType = VectorizedArray<Number>>
+class FEEvaluationBase
+  : public FEEvaluationMappingRelated<dim, Number, is_face, VectorizedArrayType>
```

Comment:
`FEEvaluation` -> `FEEvaluationAccess` -> `FEEvaluationBase` -> `FEEvaluationMappingRelated`: puh, I somewhat lost the the overview that is responsible for what (see also comment above).

Reply:
The problem with the class it that it brings together so many things. My summary would be:
- `FEEvaluation` contains the degree-specific parts of tensor products (and the user-looking class)
- `FEEvalationAccess` provides the access functionality to quantities at quadrature points
- `FEEvaluationBase` implements the vector access plus the generic part about the quadrature point access so that the `...Access` class does not need to re-implement things between different classes
- The new base class holds some generic value arrays as well as mapping-related data not directly tied to the quantities that get evaluated. So that might give us a hint to what the name could be. I could also think of `FEEvaluationDataMetric` to reflect that we have some data and metric term in this class.